nixos/modules/services/web-apps/gancio.nix

```diff
@@ -0,0 +1,253 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.gancio;
+  json = pkgs.formats.json { };
+in
+{
+  options.services.gancio = {
+    enable = lib.mkEnableOption "Gancio, a shared agenda for local communities";
+
+    package = lib.mkPackageOption pkgs "gancio" { };
+
+    plugins = lib.mkOption {
+      type = with lib.types; listOf package;
+      example = lib.literalExpression "[ pkgs.gancioPlugins.telegram-bridge ]";
+      description = ''
+        Paths of gancio plugins to activate (linked under $WorkingDirectory/plugins/).
+      '';
+    };
+
+    user = lib.mkOption {
+      type = lib.types.str;
+      description = "The user (and database name) used to run the gancio server";
+      default = "gancio";
+    };
+
+    hostName = lib.mkOption {
+      type = lib.types.str;
+      default = "localhost";
+      description = "The domain under which gancio will be reachable via nginx";
+    };
+
+    listenHost = lib.mkOption {
+      type = lib.types.str;
+      default = "localhost";
+      example = "::";
+      description = ''
+        The address (IPv4, IPv6 or DNS) for the gancio server to listen on.
+      '';
+    };
+
+    listenPort = lib.mkOption {
+      type = lib.types.port;
+      default = 13120;
+      description = ''
+        Port number of the gancio server to listen on.
+      '';
+    };
+
+    dbDialect = lib.mkOption {
+      type = lib.types.enum [
+        "sqlite"
+        "postgres"
+      ];
+      default = "sqlite";
+      description = ''
+        The database dialect to use
+      '';
+    };
+
+    logLevel = lib.mkOption {
+      description = "Gancio log level.";
+      type = lib.types.enum [
+        "debug"
+        "info"
+        "warning"
+        "error"
+      ];
+      default = "info";
+    };
+
+    settings = lib.mkOption rec {
```

Comment:
I dislike the option API that we expose here, as it imo introduces an unnecessary abstraction. The options `hostName`, `listenHost`, `listenPort`, `dbDialect` and `logLevel` all map pretty much one to one to `settings` values and should be just set as such.

- Afaik we don't really have a convention for a common webservice abstraction, e.g. `listenHost` is only exposed by 4 modules currently.
- Users of gancio have to engage with upstreams configuration format anyway, so its best to not abstract over it so much.
- We don't lose the ability to document options, assert conditions and types, require them to be set, provide defaults etc.

Have a look at `nixos/modules/services/networking/ssh/sshd.nix` for an example.

Reply:
Btw, if you agree with this but lack the time, I can write a patch to fix this.

Reply:
make sense, I'll be grateful for a patch, thank-you!

Reply:
I pushed a patch, hoping that I kept all functionality.
Explicitly setting `user_locale` and `uploads` seemed redundant, so I didn't port these options.